decision-records/2024-10-09--adding-k8s-labels-from-libs.md

```diff
@@ -0,0 +1,68 @@
+# Adding k8s Labels From Libraries
+**Date:** 2024-07-16  
+**Authors:** @dstathis @ca-scribner @IbraAoad
+
+## Context and Problem Statement
+
+To mark a pod as belonging to a given mesh, we need to apply a few labels.
+For example, with istio we would need to e.g.:
+```python
+{
+  "istio.io/dataplane-mode": "ambient",
+  "istio.io/use-waypoint": waypoint_name,
+  "istio.io/use-waypoint-namespace": waypoint_k8s_namespace,
+}
+```
+
+In order for charms to join the mesh, the beacon charm sends this set of k8s labels over relation data that the charm must apply to
+its pods. The `service_mesh` library should attempt to manage those labels.
+
+Juju manages the labels of all deployments, so charm code must re-apply the labels after key lifecycle events:
+
+1. Labels must match relation data after a leader change
+2. Labels must match relation data after a pod restart
+3. Labels must match relation data when relation data changes
+4. Labels applied by anything outside the library should remain untouched
```

Comment:
I believe the problem is simpler than we first thought.

Each unit of a charm runs in its own pod.  If we put our labels *on each Pod* (not the StatefulSet's Pod template)[^1], then each unit of a charm can be responsible for labeling itself regardless of whether it is the the leader or not.  Given that, we don't need to handle the leader change event at all (1).

If we've only labeled the Pod, Pod restarts (2) will always remove our custom labels (because the Pod will be restarted to the spec defined in the StatefulSet).  Thus we don't need to look at any StoredState in this case because **we know we cannot own any of the current labels**.  We must note though that a Pod restart **does not** trigger any relation events (I tried deleting a Pod just now and saw the following events: `[stop, upgrade_charm, config_changed, start]`).  So to handle a restart, we have to watch at least one of those events. 
  
Based on this, I'm proposing a new solution below in a separate thread that can work solely with StoredState.

[^1]: Manipulating a Pod's label does not induce a restart.  So if a charm adds a label to its own Pod, no restart is triggered.  Alternatively, we could control the labels by editing the `StatefulSet`'s `Pod` template (`StatefulSet.spec.template.metadata.labels`), adding Istio's labels there.  In this case we'd want only the leader to do this, so different units wouldn't fight over it.  Doing this, however, triggers a rolling restart so all charm units will be restarted by this change.  See the rejected solution proposed below for more details on negatives of editing the StatefulSet's template